File: services/eval/src/embedding/voyage.ts
import type { EmbeddingConfig, QueryEmbeddingClient } from "./types.js";

export class VoyageQueryEmbedding implements QueryEmbeddingClient {
  private readonly apiKey: string;
  readonly model: string;
  readonly dimensions: number;
  private readonly baseUrl = "https://api.voyageai.com/v1";

  constructor(config: EmbeddingConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model || "voyage-2";
    this.dimensions = config.dimensions ?? 1024;
  }

  async embedQuery(text: string): Promise<number[]> {
    const response = await fetch(`${this.baseUrl}/embeddings`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: this.model,
        input: [text],
        input_type: "query",
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Voyage embedding failed (${response.status}): ${error}`);
    }

    const data = (await response.json()) as {
      data: Array<{ embedding: number[] }>;
    };
    return data.data[0].embedding;
  }
}

File: services/eval/src/rag_eval/embeddings/voyage.py
from typing import Self

import httpx

from rag_eval.http_retry import post_with_retry

DEFAULT_MODEL = "voyage-2"
DEFAULT_DIMENSIONS = 1024


class VoyageQueryEmbedding:
    """Voyage query embedding. Uses ``input_type: "query"`` so the vector is
    optimized for retrieval against document-embedded chunks."""

    def __init__(
        self,
        api_key: str,
        model: str,
        dimensions: int | None = None,
        base_url: str = "https://api.voyageai.com/v1",
        client: httpx.Client | None = None,
    ) -> None:
        self._api_key = api_key
        self.model = model or DEFAULT_MODEL
        self.dimensions: int | None = dimensions if dimensions is not None else DEFAULT_DIMENSIONS
        self._base_url = base_url
        self._client = client or httpx.Client(timeout=30.0)

    def embed_query(self, text: str) -> list[float]:
        data = post_with_retry(
            self._client,
            f"{self._base_url}/embeddings",
            service="Voyage embedding",
            json={"model": self.model, "input": [text], "input_type": "query"},
            headers={"Authorization": f"Bearer {self._api_key}"},
        )
        return data["data"][0]["embedding"]

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()
